### PR TITLE
Ruby Support - More EVP_PKEY_DSA

### PR DESCRIPTION
**Pull request by justsmth**
### Issues:
Addresses CryptoAlg-2695

### Description of changes: 
Expand support for `EVP_PKEY_DSA` operations:
* `EVP_PKEY_paramgen`
  * `EVP_PKEY_CTX_set_dsa_paramgen_bits`
  * `EVP_PKEY_CTX_set_dsa_paramgen_md`
  * `EVP_PKEY_CTX_set_dsa_paramgen_q_bits`
* `EVP_PKEY_CTX_ctrl_str`
  * `"dsa_paramgen_bits"`
  * `"dsa_paramgen_md"`
  * `"dsa_paramgen_q_bits"`
* `EVP_PKEY_keygen`
* `EVP_PKEY_sign` and `EVP_DigestSign*`
* `EVP_PKEY_verify` and `EVP_DigestVerify*`

### Testing:
Tests added to cover new functionality.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: include/openssl/evp.h [hunk 1]
```diff
@@ -1289,20 +1289,29 @@ OPENSSL_EXPORT void OpenSSL_add_all_digests(void);
...
+// |EVP_PKEY_DSA| is deprecated, but signing or verifying are still supported,
+// as is parsing DER into a DSA |EVP_PKEY|.
 
 #define EVP_PKEY_DSA NID_dsa
 
-// EVP_PKEY_CTX_set_dsa_paramgen_bits returns zero.
-OPENSSL_EXPORT OPENSSL_DEPRECATED int EVP_PKEY_CTX_set_dsa_paramgen_bits(
+// EVP_PKEY_CTX_set_dsa_paramgen_bits sets the number of bits for DSA paramgen.
+// |nbits| must be larger than 256. Returns 1 on success, 0 otherwise.
+OPENSSL_EXPORT int EVP_PKEY_CTX_set_dsa_paramgen_bits(
```
Comment (WillChilds-Klein) [inline include/openssl/evp.h:1301]: should we mark these as `OPENSSL_DEPRECATED`?

File: include/openssl/evp.h [hunk 2]
```diff
@@ -1289,20 +1289,29 @@ OPENSSL_EXPORT void OpenSSL_add_all_digests(void);
...
     EVP_PKEY_CTX *ctx, int nbits);
 
-// EVP_PKEY_CTX_set_dsa_paramgen_q_bits returns zero.
-OPENSSL_EXPORT OPENSSL_DEPRECATED int EVP_PKEY_CTX_set_dsa_paramgen_q_bits(
+// EVP_PKEY_CTX_set_dsa_paramgen_md sets the digest function used for DSA
+// parameter generation. If not specified, one of SHA-1, SHA-224, or SHA-256 is
+// selected on the number of bits in |q|.
```
Comment (WillChilds-Klein) [inline include/openssl/evp.h:1306]:
```suggestion
// selected based on the number of bits in |q|.
```

File: include/openssl/evp.h [hunk 3]
```diff
@@ -1289,20 +1289,29 @@ OPENSSL_EXPORT void OpenSSL_add_all_digests(void);
...
+OPENSSL_EXPORT int EVP_PKEY_CTX_set_dsa_paramgen_md(EVP_PKEY_CTX *ctx, const EVP_MD* md);
+
+// EVP_PKEY_CTX_set_dsa_paramgen_q_bits sets the number of bits in q to use for
```
Comment (WillChilds-Klein) [inline include/openssl/evp.h:1309]: should we document here what the supported values are?

File: crypto/evp_extra/p_dsa.c [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+        case NID_sha3_512:
+          dctx->md = md;
+          return 1;
+        default:
+          OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_DIGEST_TYPE);
+          return 0;
+      }
+    }
+    case EVP_PKEY_CTRL_GET_MD:
+      *(const EVP_MD **)p2 = dctx->md;
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_dsa.c:270]: should we null-check `p2` before dereferencing?

File: crypto/evp_extra/p_dsa.c [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+      return 1;
+
+    case EVP_PKEY_CTRL_DIGESTINIT:
+    case EVP_PKEY_CTRL_PKCS7_SIGN:
+    case EVP_PKEY_CTRL_CMS_SIGN:
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_dsa.c:275]: we don't need to do anything for these cases?

File: crypto/evp_extra/p_dsa.c [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+#include <openssl/dsa.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+
+#include "../dsa/internal.h"
+#include "../internal.h"
+#include "./internal.h"
+
+typedef struct {
+  int nbits;  // defaults to 2048A
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_dsa.c:13]: nit: what does `A` mean here?

File: crypto/evp_extra/p_dsa.c [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+      case 224:
+        pmd = EVP_sha224();
+        break;
+      case 256:
+        pmd = EVP_sha256();
+        break;
+      default:
+        // This should not be possible.
+        OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_OPERATION);
+        return 0;
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_dsa.c:107]:
```suggestion
        goto end;
```

File: crypto/evp_extra/p_dsa.c [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+  GUARD_PTR(ctx->pkey);
+  GUARD_PTR(ctx->pkey->pkey.ptr);
+  GUARD_PTR(ctx->data);
+
+  DSA_PKEY_CTX *dctx = ctx->data;
+  DSA *dsa = ctx->pkey->pkey.dsa;
+
+  if (sig == NULL) {
+    // Passing NULL for sig indicates a query for the size of the signature
+    *siglen = DSA_size(dsa);
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_dsa.c:141]: should we `GUARD_PTR(siglen)` at the top?

File: crypto/evp_extra/p_dsa.c [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+        case NID_sha3_384:
+        case NID_sha3_512:
+          dctx->md = md;
+          return 1;
+        default:
+          OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_DIGEST_TYPE);
+          return 0;
+      }
+    }
+    case EVP_PKEY_CTRL_GET_MD:
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_dsa.c:269]: nit: this isn't covered in tests. do we need it?

File: crypto/evp_extra/evp_extra_test.cc [hunk 10]
```diff
@@ -3072,12 +3093,279 @@ TEST(EVPExtraTest, ParamgenCallbacks) {
...
+  {
+    bssl::UniquePtr<EVP_MD_CTX> md_ctx(EVP_MD_CTX_new());
+    ASSERT_EQ(1, EVP_DigestVerifyInit(md_ctx.get(), nullptr, EVP_sha1(), nullptr, public_key.get()));
+    ASSERT_EQ(1, EVP_DigestVerifyUpdate(md_ctx.get(), data, data_len));
+    ASSERT_EQ(1, EVP_DigestVerifyFinal(md_ctx.get(), sig.data(), sig.size()));
+  }
+
+}
+
+TEST(EVPExtraTest, DSADigestSignFinalVerify) {
```
Comment (samuel40791765) [inline crypto/evp_extra/evp_extra_test.cc:3309]: Minor nit: can we swap the `EVP_PKEY_sign_init` operations above with `EVP_DigestSignInit` operations below?

That way we'll be testing the same family of APIs: `EVP_DigestSignInit/Update/Final` with `EVP_DigestVerifyInit/Update/Final` and `EVP_PKEY_sign` with `EVP_PKEY_verify`.

File: crypto/evp_extra/p_dsa.c (hunk 9 above)
Comment (justsmth) [reply crypto/evp_extra/p_dsa.c:269]: I'm adding test coverage for it.

File: crypto/evp_extra/p_dsa.c (hunk 5 above)
Comment (justsmth) [reply crypto/evp_extra/p_dsa.c:275]: Correct.  These are no-ops.

File: crypto/evp_extra/evp_extra_test.cc [hunk 11]
```diff
@@ -3072,12 +3093,287 @@ TEST(EVPExtraTest, ParamgenCallbacks) {
...
+  // Set the generator
+  if(1 != EVP_PKEY_CTX_set_dsa_paramgen_md(ctx.get(), md)) {
+    goto dsa_paramgen_end;
+  }
+  if(1 != EVP_PKEY_paramgen(ctx.get(), &pkey_raw)) {
+    goto dsa_paramgen_end;
+  }
+  params.reset(pkey_raw);
+dsa_paramgen_end:
+  return params;
```
Comment (samuel40791765) [inline crypto/evp_extra/evp_extra_test.cc:3140]: Nit: Can we consolidate these into one if statement and just return `params` if any fail? Same for the other helper functions below as well.

Our C++ code has mostly gotten rid of `goto` statements in favor of using scoped pointers. I don't think there's a necessity to use them here.

File: crypto/evp_extra/evp_extra_test.cc (hunk 10 above)
Comment (samuel40791765) [reply crypto/evp_extra/evp_extra_test.cc:3309]: No problem, should we document that somewhere here? It's hard to tell just from looking at the test and we'll probably want to remember this intention further down the line.

File: crypto/evp_extra/p_dsa.c [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+  dctx = (DSA_PKEY_CTX *)dst->data;
+  if (sctx == NULL || dctx == NULL) {
+    return 0;
+  }
+
+  dctx->nbits = sctx->nbits;
+  dctx->qbits = sctx->qbits;
+  dctx->pmd = sctx->pmd;
+  dctx->md = sctx->md;
+  return 1;
```
Comment (WillChilds-Klein) [inline crypto/evp_extra/p_dsa.c:50]: minor nit: why not just `OPENSSL_memcpy` the whole struct? with per-field copies, if we update the struct in the future (admittedly unlikely for DSA), we'd need to remember to update this copy function too.

**Review by WillChilds-Klein — APPROVED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Pull request merged**